File: api/app/signing.py
"""Mint HMAC-signed gateway API keys.

Format (must match the Go gateway in internal/signing):
  rk_<base64url(payloadJSON)>.<base64url(hmac_sha256(payloadB64, master_key))>

The HMAC is computed over the base64-encoded payload string, so the gateway
verifies without re-serializing JSON (key order is irrelevant).
"""
from __future__ import annotations

import base64
import hashlib
import hmac
import json
import time

_PREFIX = "rk_"


def _b64url(raw: bytes) -> str:
    return base64.urlsafe_b64encode(raw).rstrip(b"=").decode()


def _mac(master_key: str, msg: str) -> bytes:
    return hmac.new(master_key.encode(), msg.encode(), hashlib.sha256).digest()


def sign(
    master_key: str,
    user_id: str,
    key_id: str,
    provider_id: str = "",
    issued_at: int | None = None,
) -> str:
    """Produce a signed gateway key for the given claims."""
    claims: dict[str, object] = {
        "uid": user_id,
        "kid": key_id,
        "iat": issued_at if issued_at is not None else int(time.time()),
    }
    if provider_id:
        claims["pid"] = provider_id
    payload = json.dumps(claims, separators=(",", ":")).encode()
    enc_payload = _b64url(payload)
    sig = _mac(master_key, enc_payload)
    return _PREFIX + enc_payload + "." + _b64url(sig)

File: gateway/internal/signing/signing.go
// Package signing implements zero-network API key auth.
//
// A gateway key is a self-describing, HMAC-SHA256 signed token:
//
//	rk_<base64url(payloadJSON)>.<base64url(hmac)>
//
// The gateway verifies it with CPU math only — no database round-trip.
// The identical scheme is implemented in the Python api (app/signing.py)
// so the api can mint keys the gateway trusts.
package signing

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"errors"
	"strings"
)

const prefix = "rk_"

// ErrInvalidKey is returned when a token is malformed or its signature fails.
var ErrInvalidKey = errors.New("invalid api key")

// Claims is the payload embedded in a signed key.
type Claims struct {
	UserID     string `json:"uid"`
	KeyID      string `json:"kid"`
	IssuedAt   int64  `json:"iat"`
	ProviderID string `json:"pid,omitempty"`
}

// Sign produces a signed key for the given claims using the master key.
func Sign(masterKey string, c Claims) (string, error) {
	payload, err := json.Marshal(c)
	if err != nil {
		return "", err
	}
	encPayload := base64.RawURLEncoding.EncodeToString(payload)
	sig := mac(masterKey, encPayload)
	return prefix + encPayload + "." + base64.RawURLEncoding.EncodeToString(sig), nil
}

// Verify validates a signed key and returns its claims.
// It performs a constant-time signature comparison and never touches the DB.
func Verify(masterKey, token string) (Claims, error) {
	var c Claims
	if !strings.HasPrefix(token, prefix) {
		return c, ErrInvalidKey
	}
	body := token[len(prefix):]
	dot := strings.IndexByte(body, '.')
	if dot < 0 {
		return c, ErrInvalidKey
	}
	encPayload, encSig := body[:dot], body[dot+1:]

	gotSig, err := base64.RawURLEncoding.DecodeString(encSig)
	if err != nil {
		return c, ErrInvalidKey
	}
	wantSig := mac(masterKey, encPayload)
	if !hmac.Equal(gotSig, wantSig) {
		return c, ErrInvalidKey
	}

	payload, err := base64.RawURLEncoding.DecodeString(encPayload)
	if err != nil {
		return c, ErrInvalidKey
	}
	if err := json.Unmarshal(payload, &c); err != nil {
		return c, ErrInvalidKey
	}
	return c, nil
}

func mac(masterKey, msg string) []byte {
	h := hmac.New(sha256.New, []byte(masterKey))
	h.Write([]byte(msg))
	return h.Sum(nil)
}
